### ISSUE TITLE
The spark cache is too large

### ISSUE THREAD

**Issue by mingnet** [locationtech/geotrellis#2819]
I want to know how much memory is needed when the geotrellis runs, so I set the cache policy  of spark to `DISK_ONLY`. I use geotrellis2.0.0 and spark 2.3.0, I have 10G data and the spark has 3 working nodes. The data is stored in hdfs, I am trying to load geotiff data through etl.
According to the documentation (https://docs.geotrellis.io/en/latest/tutorials/etl-tutorial.html), I compile the geotrellis-spark-etl-assembly-2.0.0-SNAPSHOT.jar file.
Then I create the input.json file.

```
[{
    "format": "multiband-geotiff",
    "name": "china",
    "cache": "DISK_ONLY",
    "backend": {
        "type": "hadoop",
        "path": "hdfs://192.168.1.41:9000/tiletest/201705"
    }
}]
```

Output.json file
```
{
    "backend": {
        "type": "hadoop",
        "path": "hdfs://192.168.1.41:9000/result/test/"
    },
    "reprojectMethod": "buffered",
    "pyramid": true,
    "tileSize": 256,
    "keyIndexMethod": {
        "type": "zorder"
    },
    "resampleMethod": "nearest-neighbor",
    "layoutScheme": "zoomed",
    "crs": "EPSG:3857"
}

```

Backend-profiles.json file

```
{
    "backend-profiles": []
}
```

Then I execute the program through the following command
```sh
/opt/spark/bin/spark-submit --class geotrellis.spark.etl.MultibandIngest --master 'spark://192.168.1.8:7077' --driver-memory 50G /work/etl/geotrellis-spark-etl -assembly-2.0.0-SNAPSHOT.jar --input "file:///work/etl/input.json" --output "file:///work/etl/output.json" --backend-profiles " File:///work/etl/backend-profiles.json"
```

Then I found that the cache of the three working nodes of spark accounted for 25G of disk space.The cache per node is 2.5 times as much the original data.

Follow this. If I ingest 100G or 1TB of data, it will require a lot of cache space. I hope I can use less space so I can just use memory.

I want to know where there is something wrong, what do I need to do to improve it.

**Comment by pomadchin** [comment]
Hey @mingnet what GeoTrellis version do you use?

**Comment by mingnet** [comment]
Hey @pomadchin  I need to correct the value of my question. Because I ignored the compression option for tiff data. The above value is the geotiff data size processed with gdal_retile.py, using the `-co compress=deflate` option. 

I tried to turn off this option. The resources used in the process are as follows:
24G tiff data, each working node  of spark   accounted for 25G of disk space. 
8G tiff data, each working node  of spark  accounted for 10G of disk space.

In other words, each node must be configured with at least the same amount of memory or disk space as the data. I don't think this is very reasonable. These data should be evenly distributed to 3 nodes. Even use fewer resources, because it should be possible to generate only one piece of data for the target grid at a time.

Suppose we need to process 1T data. It will be very difficult.

the GeoTrellis version is 2.0.0. 

gdalinfo(before gdal_retile.py processing):

```shell
Driver: GTiff/GeoTIFF
Files: /data/remotesensing/201704/E116D8_N36D0_20170417_GF1_WFV2_DOM.tif
Size is 18315, 15387
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.2572235604902,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Origin = (115.441238182399985,37.067668339199997)
Pixel Size = (0.000144000000000,-0.000144000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 115.4412382,  37.0676683) (115d26'28.46"E, 37d 4' 3.61"N)
Lower Left  ( 115.4412382,  34.8519403) (115d26'28.46"E, 34d51' 6.99"N)
Upper Right ( 118.0785982,  37.0676683) (118d 4'42.95"E, 37d 4' 3.61"N)
Lower Right ( 118.0785982,  34.8519403) (118d 4'42.95"E, 34d51' 6.99"N)
Center      ( 116.7599182,  35.9598043) (116d45'35.71"E, 35d57'35.30"N)
Band 1 Block=18315x1 Type=UInt16, ColorInterp=Gray
Band 2 Block=18315x1 Type=UInt16, ColorInterp=Undefined
Band 3 Block=18315x1 Type=UInt16, ColorInterp=Undefined
Band 4 Block=18315x1 Type=UInt16, ColorInterp=Undefined
Driver: GTiff/GeoTIFF
Files: /data/remotesensing/201704/E114D2_N34D7_20170417_GF1_WFV1_DOM.tif
Size is 21192, 15671
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.2572235604902,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Origin = (112.548738298100005,35.793564391800004)
Pixel Size = (0.000144000000000,-0.000144000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 112.5487383,  35.7935644) (112d32'55.46"E, 35d47'36.83"N)
Lower Left  ( 112.5487383,  33.5369404) (112d32'55.46"E, 33d32'12.99"N)
Upper Right ( 115.6003863,  35.7935644) (115d36' 1.39"E, 35d47'36.83"N)
Lower Right ( 115.6003863,  33.5369404) (115d36' 1.39"E, 33d32'12.99"N)
Center      ( 114.0745623,  34.6652524) (114d 4'28.42"E, 34d39'54.91"N)
Band 1 Block=21192x1 Type=UInt16, ColorInterp=Gray
Band 2 Block=21192x1 Type=UInt16, ColorInterp=Undefined
Band 3 Block=21192x1 Type=UInt16, ColorInterp=Undefined
Band 4 Block=21192x1 Type=UInt16, ColorInterp=Undefined
Driver: GTiff/GeoTIFF
Files: /data/remotesensing/201704/E116D3_N34D3_20170417_GF1_WFV2_DOM.tif
Size is 17916, 15376
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.2572235604902,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Origin = (115.032904865400013,35.398584405899996)
Pixel Size = (0.000144000000000,-0.000144000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 115.0329049,  35.3985844) (115d 1'58.46"E, 35d23'54.90"N)
Lower Left  ( 115.0329049,  33.1844404) (115d 1'58.46"E, 33d11' 3.99"N)
Upper Right ( 117.6128089,  35.3985844) (117d36'46.11"E, 35d23'54.90"N)
Lower Right ( 117.6128089,  33.1844404) (117d36'46.11"E, 33d11' 3.99"N)
Center      ( 116.3228569,  34.2915124) (116d19'22.28"E, 34d17'29.44"N)
Band 1 Block=17916x1 Type=UInt16, ColorInterp=Gray
Band 2 Block=17916x1 Type=UInt16, ColorInterp=Undefined
Band 3 Block=17916x1 Type=UInt16, ColorInterp=Undefined
Band 4 Block=17916x1 Type=UInt16, ColorInterp=Undefined
Driver: GTiff/GeoTIFF
Files: /data/remotesensing/201704/E114D6_N36D3_20170417_GF1_WFV1_DOM.tif
Size is 21660, 15665
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.2572235604902,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Origin = (112.903738283899969,37.463533658299994)
Pixel Size = (0.000144000000000,-0.000144000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 112.9037383,  37.4635337) (112d54'13.46"E, 37d27'48.72"N)
Lower Left  ( 112.9037383,  35.2077737) (112d54'13.46"E, 35d12'27.99"N)
Upper Right ( 116.0227783,  37.4635337) (116d 1'22.00"E, 37d27'48.72"N)
Lower Right ( 116.0227783,  35.2077737) (116d 1'22.00"E, 35d12'27.99"N)
Center      ( 114.4632583,  36.3356537) (114d27'47.73"E, 36d20' 8.35"N)
Band 1 Block=21660x1 Type=UInt16, ColorInterp=Gray
Band 2 Block=21660x1 Type=UInt16, ColorInterp=Undefined
Band 3 Block=21660x1 Type=UInt16, ColorInterp=Undefined
Band 4 Block=21660x1 Type=UInt16, ColorInterp=Undefined
```

**Comment by pomadchin** [comment]
All right, that's what i thought. The reason you need such heavy nodes is because you have a tiff with 15665 segments, each of this segment is a strip sized 21660. It means to chunk this tiff by squares requires a lot of data. I'll prepare a bit more expanded answer a lil bit later, but the main problem here is that it's very hard to read you tiff windowed. Also taking into account that you used a deflate compression to compress the source tiff, it means a rather heavy CPU load in this case during the ingest process just to read data into Spark mem.

**Comment by mingnet** [comment]
I will try your suggestion.

However, I have some problems. If set to 256, a large amount of data will generate a large number of files, so the indexing efficiency of the file system will be very low. I usually retile tiff to square segments (8192x8192). 

What is your opinion about this?

**Comment by pomadchin** [comment]
@mingnet ah, just to be clear, i talk about _segment layout tiling_, it will create no extra files - just a single tiff, though it's internal segments representation would be different. Try this:

```
gdal_translate -co "TILED=YES" input.tiff output.tiff
```

and after that look into `gdalinfo` of the output tiff. You'll see that the blocksize of the tiff would be `256x256`

**Comment by mingnet** [comment]
@pomadchin thanks. I know how to do it.

**Comment by pomadchin** [comment]
@mingnet hm, so what do you mean by `a large number of files`?

**Comment by mingnet** [comment]
@pomadchin very, very many small files. Maybe millions.  I thought that each segment is divided into a tiff file.

**Comment by pomadchin** [comment]
@mingnet gotcha! so each tiff is contained of segments that allows us to read them in a streaming fashion.

The way we read tiffs into memory is smth like solving a bin packing problem. We're trying to pack tiff segments into spatially co-located windows (by default sized 256x256), such windows would represent a partition limited by 128mb by default. In this case our partitioning scheme would be a set of windows distributed across the Spark cluster. Such approach does not require to read all the tiff into memory just to read a small window.
During the read you receive `N` partitions, filled with `(Extent, Tile)` pairs, where `Tile` in the best case is just a `tiff segment`. It allows to create a cheapest and a well distributed RDD partitioning scheme.

If your tiff contains large segments, or large strip segments (like in your case), we'll try to pack large segments into partitions, and after that we'll have to shuffle a lot of data to crop out a square window on the next step, during creation of a tiled layout (obviously, strips are not square and to receive a square window we need to read the entire set of strip segments to crop them into memory of a node).

**Comment by mingnet** [comment]
@pomadchin I follow what you said. But still the same as before.
I don't understand why this is happening. I will upload the data and files to aws's s3. Can you use my data to check this?
I am uploading data and will be able to complete it tomorrow. If you are willing to test this, I will send you the aws s3 address.

**Comment by pomadchin** [comment]
@mingnet sounds good :+1: also a followup question, can you throw gdalinfo of the new generated file?

**Comment by mingnet** [comment]
@pomadchin I have uploaded data to aws s3. The address is s3://gtdatatest. This directory contains all files. You can download and test it. 

gdalinfo(One of the tiff files)

Driver: GTiff/GeoTIFF
Files: /data1/data20181031/a_r_1_1.tif
Size is 8192, 8192
Coordinate System is:
GEOGCS["WGS 84",
    DATUM["WGS_1984",
        SPHEROID["WGS 84",6378137,298.257223563,
            AUTHORITY["EPSG","7030"]],
        AUTHORITY["EPSG","6326"]],
    PRIMEM["Greenwich",0],
    UNIT["degree",0.0174532925199433],
    AUTHORITY["EPSG","4326"]]
Origin = (113.533894000000004,-26.867094999999999)
Pixel Size = (0.000144000000000,-0.000144000000000)
Metadata:
  AREA_OR_POINT=Area
Image Structure Metadata:
  INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left  ( 113.5338940, -26.8670950) (113d32' 2.02"E, 26d52' 1.54"S)
Lower Left  ( 113.5338940, -28.0467430) (113d32' 2.02"E, 28d 2'48.27"S)
Upper Right ( 114.7135420, -26.8670950) (114d42'48.75"E, 26d52' 1.54"S)
Lower Right ( 114.7135420, -28.0467430) (114d42'48.75"E, 28d 2'48.27"S)
Center      ( 114.1237180, -27.4569190) (114d 7'25.38"E, 27d27'24.91"S)
Band 1 Block=256x256 Type=UInt16, ColorInterp=Gray
Band 2 Block=256x256 Type=UInt16, ColorInterp=Undefined
Band 3 Block=256x256 Type=UInt16, ColorInterp=Undefined
Band 4 Block=256x256 Type=UInt16, ColorInterp=Undefined

**Comment by pomadchin** [comment]
Btw @mingnet have you tried to disable caching? `"cache": "NONE"`

**Comment by mingnet** [comment]
@pomadchin  I just tested this and found that it was still the same as before.

**Comment by pomadchin** [comment]
Yea @mingnet thx for that; this sprint we scheduled to look into it! Thx!

**Comment by mingnet** [comment]
@pomadchin ok. Now  you can access to my bucket. s3://gtdatatest

**Comment by pomadchin** [comment]
Ok, i can list objects at the moment, but can't download them into a local disk / or to copy into our bucket. `when calling the {UploadPartCopy | GetObject} operation: Access Denied`

**Comment by mingnet** [comment]
sorry. I am not proficient in s3. let me try again.

**Comment by mingnet** [comment]
command:
spark-submit --class geotrellis.spark.etl.MultibandIngest --master 'spark://192.168.1.8:7077' --driver-memory 50G /work/etl/geotrellis-spark-etl-assembly-2.1.0.jar --input "file:///work/etl/input.json" --output "file:///work/etl/output.json" --backend-profiles "file:///work/etl/backend-profiles.json"

The spark deployed in standalone mode.  It has 3 working nodes. This spark is spark-2.3.0-bin-hadoop2.7. 

spark-defaults.conf
```shell
spark.executor.extraJavaOptions -Xss32m
spark.driver.extraJavaOptions=-Xss32m

PYSPARK_PYTHON=/opt/python3.7
```

spark-env.sh
```
export JAVA_HOME=/usr/lib/jvm/jre

export SCALA_HOME=/opt/scala

export HADOOP_HOME=/opt/hadoop

export HADOOP_CONF_DIR=/opt/hadoop/etc/hadoop

export SPARK_MASTER_HOST=master

export SPARK_MASTER_PORT=7077

export SPARK_LOCAL_DIRS=/data1

export SPARK_WORKER_MEMORY=50g

export SPARK_EXECUTOR_CORES=3
export SPARK_CORES_MAX=6

export SPARK_DRIVER_MEMORY=50g
export SPARK_EXECUTOR_MEMORY=50g
export PYSPARK_PYTHON=/opt/python3.7/bin/python3
```

**Comment by pomadchin** [comment]
Ok, wondering do you see that data is distributed across the cluster in the UI? (just a double check to confirm the behaviour)

**Comment by mingnet** [comment]
I don't quite understand which one you are talking about. Is it the following data?

```shell

ExecutorID | Worker | Cores | Memory | State | Logs
11 | worker-20181019165435-192.168.1.165-42882 | 3 | 51200 | EXITED | stdout stderr
1 | worker-20181019165435-192.168.1.165-42882 | 3 | 51200 | EXITED | stdout stderr
3 | worker-20181102141149-192.168.1.165-45754 | 3 | 51200 | EXITED | stdout stderr
17 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
2 | worker-20181019165432-192.168.1.8-46594 | 3 | 51200 | EXITED | stdout stderr
13 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
6 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
16 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
9 | worker-20181102141149-192.168.1.35-36166 | 3 | 51200 | EXITED | stdout stderr
12 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
4 | worker-20181102141149-192.168.1.35-36166 | 3 | 51200 | EXITED | stdout stderr
0 | worker-20181102141149-192.168.1.8-36191 | 3 | 51200 | EXITED | stdout stderr
8 | worker-20181019165432-192.168.1.8-46594 | 3 | 51200 | EXITED | stdout stderr



```

**Comment by pomadchin** [comment]
Kind of; though requested just a web ui page (`spark-master:4040`).

**Comment by pomadchin** [comment]
I ran an ingest on the `a_r_1_1.tif` tiff: 

![image](https://user-images.githubusercontent.com/4929546/48627642-22b6c580-e9c6-11e8-985c-aa89a29d6ae9.png)

![image](https://user-images.githubusercontent.com/4929546/48627695-4548de80-e9c6-11e8-83b1-9d985e9abff8.png)

Config is the same as yours, i ingested from s3 on s3.
Launch options: 

```
spark-submit,--master,yarn-cluster,\
--class,geotrellis.spark.etl.MultibandIngest,\
--driver-memory,4200M,\
--driver-cores,2,\
--executor-memory,4200M,\
--executor-cores,2,\
--conf,spark.dynamicAllocation.enabled=true,\
--conf,spark.yarn.executor.memoryOverhead=700,\
--conf,spark.yarn.driver.memoryOverhead=700,\
s3://geotrellis-test/daunnc/geotrellis-spark-etl-assembly-${GEOTRELLIS_VERSION}${GEOTRELLIS_VERSION_SUFFIX}.jar,\
--input,"s3://geotrellis-test/issue-2819/cfg/input.json",\
--output,"s3://geotrellis-test/issue-2819/cfg/output.json",\
--backend-profiles,"s3://geotrellis-test/issue-2819/cfg/backend-profiles.json"
```

**Comment by pomadchin** [comment]
All right, i noticed that: 

![image](https://user-images.githubusercontent.com/4929546/48629543-f3ef1e00-e9ca-11e8-8adf-b03270c42e3a.png)

**Comment by pomadchin** [comment]
Hey @mingnet, I added more nodes (10x m3.xlarge nodes), and measured the DISK space usage:

![image](https://user-images.githubusercontent.com/4929546/48706328-f80a7e00-ec0c-11e8-9b7b-2624e0dc31bd.png)

Cluster disk usage for this task:

![image](https://user-images.githubusercontent.com/4929546/48706429-4d468f80-ec0d-11e8-9016-90b20a75d773.png)

![image](https://user-images.githubusercontent.com/4929546/48706440-57688e00-ec0d-11e8-8c1b-b836268d882f.png)

![image](https://user-images.githubusercontent.com/4929546/48706453-651e1380-ec0d-11e8-82bd-8fe7c0d26e5d.png)

![image](https://user-images.githubusercontent.com/4929546/48706465-6d764e80-ec0d-11e8-86d0-c707a3a6ab08.png)

![image](https://user-images.githubusercontent.com/4929546/48706474-75ce8980-ec0d-11e8-9bb1-2a6cc0617111.png)

![image](https://user-images.githubusercontent.com/4929546/48706488-7cf59780-ec0d-11e8-93a3-cbca935d8b33.png)

![image](https://user-images.githubusercontent.com/4929546/48706499-84b53c00-ec0d-11e8-897f-7433dd09a56d.png)

![image](https://user-images.githubusercontent.com/4929546/48706511-8da60d80-ec0d-11e8-934a-d1ee64dc6517.png)

![image](https://user-images.githubusercontent.com/4929546/48706521-94cd1b80-ec0d-11e8-9768-6b44b8d4dee9.png)

![image](https://user-images.githubusercontent.com/4929546/48706533-9e568380-ec0d-11e8-9fa0-22421633eb37.png)

So the result is that it's an expected thing and smth constant across the entire cluster. Disk space problems can be resolved by adding more nodes, and it doesn't mean that each node should have a disk space size equal to the input data size.

I'd recommend you to test it more, actually we didn't have issues on ingesting large datasets :/ 

Also It's only a temporary disk space occupation required for the saprk work. Along with that don't forget to clean up spark work dirs manually after finishing your job.

Let me know if this explanation is not good enough, and feel free to close / keep this issue open.
I'm removing a `bug` label since it doesn't seem to be a bug.

**Comment by pomadchin** [comment]
Hey @mingnet I'm closing this issue, but feel free to re-open it again if the are any more issues or you can create a new one. :+1:
